### ISSUE TITLE
Ridge Racer 1 and 2 speedometers do not show correctly

### ISSUE THREAD

**Issue by unknownbrackets** [hrydgard/ppsspp#8509]
Shortly after drawing with texture 0x09b41fb0, the game renders to the framebuffer 001e4e40.  It draws line primitives based on your speed, with matching alpha values for how filled the speedometer should be.

Next, it loads that framebuffer as a CLUT.  As of writing, this doesn't work correctly - because of a small bug in the CLUT download:

``` c++
        int w = std::min(pixels % vfb->fb_stride, (int)vfb->width);
        int h = std::min((pixels + vfb->fb_stride - 1) / vfb->fb_stride, (int)vfb->height);
```

Needs to have:

``` c++
        if (w == 0 || h > 1) {
            // Exactly aligned, or more than one row.
            w = std::min(vfb->fb_stride, (int)vfb->width);
        }
```

Anyway, it's just as well, since that would still draw incorrectly.  Anyway, it uses this CLUT now to render the speedometer.

So, what goes wrong?  Lines.  Lines go wrong.  Without #8276, the lines (at ALL resolutions) are drawn incorrectly, and so the framebuffer is not right.  It somewhat works to increase the line width, although this is still somewhat wrong.

The reason is, it draw lines e.g. from (0,0)-(5,0).  OpenGL centers the line, but the game clearly expects (this is throughmode btw) the line to be drawn downward, and this problem only gets worse the higher the internal resolution is.  But it even doesn't work at 1x for this reason.

Anyway, there's an extra problem.  Suppose that we fixed all the above - the speedometer would then render perfectly at 1x.  But... it would not render correctly at 2x.  Why?  Because of linear filtering.  Here's what happens:
1. The framebuffer is detected as being 256x272.  I'm not sure if parts of it are or are not used for other drawing, so this might be a correct estimate.
2. For the purposes of the CLUT, only 256x1 is useful.
3. At some point or another, white gets loaded in the row below the CLUT (and neither it nor the area that ought to be transparent seem to be redrawn regularly.)
4. At some point, this white ends up blended (partially) into the framebuffer.  The effect is better at >2x.  I'm not sure why, since I thought we blitted using NEAREST...

Ultimately, this means there's a white box around the speedometer.  Lines aren't the issue here, since it's not even drawing to the part of the palette that is supposed to be transparent and ends up blended with white.

So to summarize:
- Render to CLUT bug
- Line drawing bug
- > 1x PSP resolution bug

-[Unknown]

**Comment by unknownbrackets** [comment]
Just noting, the linear stuff is from the initial upload of course, and using nearest for `BufFilter` fixes it.  I guess we could make memcpy uploads use nearest by default or using some heuristics....

-[Unknown]

**Comment by spring568** [comment]
Thank you for looking into this. I Hope you can look at this again when you have free time. Also probably not terribly helpful to you, but here's a video of me playing horribly with the retail UMD on a psp 1000:  https://www.youtube.com/watch?v=bJHAb5s1C5Y

**Comment by amurgshere** [comment]
Would be great if we could #8276 and this one fixed so that the Speedos in Ridge Ricer 1 and 2 render properly, +1 for this to be fixed :) Please...

**Comment by Back2Life888** [comment]
Any kind of workarounds in here??

**Comment by hrydgard** [comment]
No, not at this time.

**Comment by Veritas83** [comment]
I think i found a workaround... I just need to confirm it now.

[img]https://i.gyazo.com/f3f36fa3aea13a49d93b8ce6dc453c6c.png[/img]

**Comment by unknownbrackets** [comment]
As I recall, more is needed, but #15076 should hopefully at least improve one of the steps here that were broken before.

-[Unknown]

**Comment by Back2Life888** [comment]
> As I recall, more is needed, but #15076 should hopefully at least improve one of the steps here **_that were broken before_**.
> 
> -[Unknown]

So you're saying that the speedometers are sort of but not actually working now?

**Comment by unknownbrackets** [comment]
> So, what goes wrong? Lines. Lines go wrong. Without #8276, the lines (at ALL resolutions) are drawn incorrectly, and so the framebuffer is not right. It somewhat works to increase the line width, although this is still somewhat wrong.

This part from the above should be fixed now.

I think the render-to-clut fix might've been reverted, but I'll try to mess with this again at some point.  With the line thing fixed, it should be more easily fixable...

-[Unknown]

**Comment by Back2Life888** [comment]
can you send a screenshot of what the speedometer looks like with #15076?

**Comment by Back2Life888** [comment]
What tool do you use to mess around with the speedometers?

**Comment by unknownbrackets** [comment]
Software rendering does seem to mostly be getting these right.

-[Unknown]

**Comment by Back2Life888** [comment]
> Software rendering does seem to mostly be getting these right.
> 
> -[Unknown]

Yes it does although it is extremely slow. You could look into how the game loads in the speedometers in Software rendering and then recreate it with the Software rendering disabled.

**Comment by unknownbrackets** [comment]
Well, I did describe how it renders speedometers above.  Software rendering is much faster in the latest git builds, though, just to note.

-[Unknown]

**Comment by Back2Life888** [comment]
I manage to run it at 25-60 FPS with Software rendering mode. The image is so weirdly crisped. Some triangles flicker around the race track. But other than that, the speedometers work correctly. Although I noticed some imperfections, specifically the Gnade speedometer. I manage to get 60 FPS on the tunnel but the Gnade speedometer is oddly buffery for some reason.
![image](https://user-images.githubusercontent.com/76780043/155078202-18e0c4cc-ea1f-4569-be36-84047f6f0be0.png)

**Comment by mrwallace888** [comment]
Still an issue, huh?

**Comment by Back2Life888** [comment]
> Still an issue, huh?

Can't read or back track before posting, huh?

**Comment by combo-pizza** [comment]
The speedometers work now, but Interestingly an error with the software render shows up when enabling clut download for ridge racer. In midtown expressway instead of there being a black triangle explosion at the beginning of the race there's instead missing polygons on the track in a few places. Another thing that happens is the headlight coronas disappear which doesn't happen in software.

**Comment by unknownbrackets** [comment]
Is the headlight issue only present on hardware renderers, or did you mean to say those were all issues in software?  It might belong in a separate issue, although very possible headlights are handled this same way as "lens flares."

-[Unknown]

**Comment by combo-pizza** [comment]
Yes I mean only in the hardware render for the headlights

**Comment by combo-pizza** [comment]
Sorry about that last post, anyway polygons was with hardware render with clut download enabled. The headlight issue has nothing to do with this

**Comment by Back2Life888** [comment]
I'm using Build 2022-09-12 08:35:13 from Buildbot. Issue still persists. However, something is different with the speedometers now. If you load up a save state of the game in Software Rendering mode with the cars with the broken speedometers, the speedometers kinda bug out in a weird way.

**Comment by combo-pizza** [comment]
Go to any level with a sun effect and it'll draw garbage all over the screen. Sunset drive is a perfect example. Looks like it's pointing to the wrong texture when drawing (some kind of memory corruption?). Probably was always there but is showing now because of clut downloads.

![image](https://user-images.githubusercontent.com/95991264/190666678-af4a781c-5f20-4ff4-9bb7-896a31ab064c.png)

**Comment by hrydgard** [comment]
Thanks for the report! That's a good clue for investigation.

**Comment by combo-pizza** [comment]
I don't think you need any hints. The sun shaft effect is broken lol and it's pretty obvious. I wish I was better at this bug report stuff

**Comment by amurgshere** [comment]
That pattern in the sun effect looks like a zoomed version of the car tachometer…

**Comment by hrydgard** [comment]
Yeah, kinda. Clearly some confusion between framebuffer binding or copies or similar. Should be solvable with a bit of investigative work...

**Comment by Back2Life888** [comment]
> Go to any level with a sun effect and it'll draw garbage all over the screen. Sunset drive is a perfect example. Looks like it's pointing to the wrong texture when drawing (some kind of memory corruption?). Probably was always there but is showing now because of clut downloads.
> 
> ![image](https://user-images.githubusercontent.com/95991264/190666678-af4a781c-5f20-4ff4-9bb7-896a31ab064c.png)
> 
> Edit: Oh now i get it, those tracks with the weird lines on the screen is some kind of effect that's not being drawn correctly in the software GE and now it's showing up in HW. I don't think this is related to the speedometer though.

Hey Combo, do you happen to be using an NVIDIA card? I'm noticing a pattern on where AMD cards and NVIDIA cards have different results and stuff.

**Comment by Back2Life888** [comment]
Updated to the newest build. The speedometers are fixed in Hardware and the weird garbagy texture thing seems to depend on your camera angle. If you face your car somewhere else in the road, it seems to disappear all of a sudden. And the effect varies depending on the course, the car you are driving, and the opponents. I noticed that when I took the Crinale for a spin in 765, the garbage that gets displayed is actually the wheel textures of the car and after getting a little bit farther from the course, it changed to a course texture instead.

**Comment by combo-pizza** [comment]
Yup that's pretty much what I'm getting too. I failed to notice it since I mainly use vulkan, but opengl does work fine

**Comment by Back2Life888** [comment]
> Yup that's pretty much what I'm getting too. I failed to notice it since I mainly use vulkan, but opengl does work fine

Yeah but OpenGL now has a big downside. It now lags when the game is rendering the speedometers that was broken before hand and it lags too when you drive in parts of some courses where the garbage stuff usually appears on Vulkan. I dunno if you are experiencing this too, I have quite the garbo lappy.

**Comment by combo-pizza** [comment]
Really? I don't seem to have any of those slowdowns, but my system is probably over the specs that it'd require. I have a 3060Ti and a  Core i5-11600K.  Hrm but I do have a core 2 duo system with a GTX 745 I pulled from a dead dell I could test it out on.

**Comment by hrydgard** [comment]
Hm, I tried Sunset Drive with the "Good luck bayonet" car, which has a CLUT-driven speedometer. To me everything seems to work. Here's the CLUT speedometer in renderdoc, working nicely:

![image](https://user-images.githubusercontent.com/130929/191355816-9072b992-a8e8-4db4-9b18-bbb49585c01e.png)

No glitches:

![image](https://user-images.githubusercontent.com/130929/191356213-9d03eeef-d7e9-4439-8458-4b008d45ad55.png)

This is Vulkan on NV.

Headlight glare is indeed broken, no idea how that works...

**Comment by combo-pizza** [comment]
If I run IR interpreter with fast memory disabled the error doesn't show up in vulkan I found out

**Comment by hrydgard** [comment]
I don't think something like resizable bar could affect it. Maybe some setting inside of PPSSPP. I will try a bit more to reproduce it, or maybe you can catch a GE dump when it happens, and see if it repros through that? https://github.com/hrydgard/ppsspp/wiki/How-to-create-a-frame-dump

**Comment by combo-pizza** [comment]
I tried getting a framedump of it and it didn't show up at all when I loaded it.
[ULUS10001.zip](https://github.com/hrydgard/ppsspp/files/9611812/ULUS10001.zip)

**Comment by unknownbrackets** [comment]
[RenderDoc](https://renderdoc.org/) is a debugger that can record frames and show you all the steps.  It also comes with the Vulkan SDK.

-[Unknown]

**Comment by combo-pizza** [comment]
Thank you! I did a capture of it, not sure if I did it correctly though
[ridgeracer.zip](https://github.com/hrydgard/ppsspp/files/9613387/ridgeracer.zip)

**Comment by hrydgard** [comment]
Just can't get it to happen, this is what I'm doing:

* Starting the euro version of Ridge Racer 1
* Down, Single Race
* Sunsite Drive
* Class 1
* Car: Good Luck Bayonet
* Start
* Drive to the sunset

No sun glow glitch, neither in debug nor release builds.

However! Comparing to software (which has some weird horizontal glitches, but probably irrelevant to this), it's clear that a sun glare sprite is missing. Presumably that's what's getting drawn for you, for some reason.

The missing sun glare sprite is probably related to the missing car spotlight glares.. I made #15923 for tracking investigation of various lens flare effects.

**Comment by combo-pizza** [comment]
Wait! I finally got it to show up in a frame dump!! here! It doesn't look the same but the error is clearly there
[ULUS10001 (1).zip](https://github.com/hrydgard/ppsspp/files/9615634/ULUS10001.1.zip)

**Comment by stuken** [comment]
I can replicate the ghost images and a bugged speedo.
Ridge racer 2, arcade mode, first track, first class 6 car.
Using 4x internal on vulkan with no texture upscaling. Adding either cpu or shader uscaling the speedo will be corrupted.
On a 3090 with the latest drivers.
![UCES00422_00001](https://user-images.githubusercontent.com/8155719/191469192-f961845b-5a8e-4fdb-848f-635e12f5aabd.jpg)
![UCES00422_00003](https://user-images.githubusercontent.com/8155719/191469716-4d31e189-c885-4b6e-9fd1-63452f82cee5.jpg)
![UCES00422_00004](https://user-images.githubusercontent.com/8155719/191470361-904f0bca-bc53-4797-bec0-8f13d74a35f6.jpg)

**Comment by combo-pizza** [comment]
It still did the same thing even when I removed the save. Not to speculate but it looks like a problem with vulkan and 30XX cards

**Comment by combo-pizza** [comment]
Yeah it still does the same thing even when I remove saves so I have no idea. I can't get a proper capture of it either

**Comment by hrydgard** [comment]
Ooh, texture upscaling seems to cause trouble here, need to look into that. There does seem to be another non-upscaling related issue too, you're not using it right, combo-pizza?

**Comment by combo-pizza** [comment]
Hrm it still happens at 1x scaling but it does change when you turn it up to 3x or 4x so you could be right about that

**Comment by combo-pizza** [comment]
Whoops,  I didn't notice you were talking about texture scaling. Yeah I'm not using any kind of upscaling at all. I also noticed some weird stuff too when I changed the internal resolution too when the sunshaft glitch was happening, but that's probably not important either.

**Comment by hrydgard** [comment]
#16076 should fix the glitches when texture scaling is enabled.

Before this change, I once managed to get a corrupted speedomeeter, with another texture applied, similar to the described glitch, but having a hard time reproducing it reliably...

It's got to be some state clearing/dirtying issue, I think.

**Comment by stuken** [comment]
Looks like it's nearly there.
Auto max quality texture filtering on vulkan is also triggering corruption. Auto and lower appear unaffected.
There is also the grey box appearing on that specific cars speedo at above native res.

**Comment by Back2Life888** [comment]
I'm kinda getting confused as to why the headlight glare broke down when the speedometers are fixed. In old builds, the headlight glare works on AMD but is bugged on NVIDIA. In NVIDIA, chances are the headlight glare would only render on 1 headlight and will render at a very short distance or won't even render at all.

**Comment by hrydgard** [comment]
That should take care of that, at least.

**Comment by hrydgard** [comment]
So I wanted to look into the headlight glare issue, and am playing on hardware, but struggling to find a single place in the game where I can see them. Not in the camera zoom around the car before the race, not in an actual race (you see the car from the back the whole time) and opponents don't have headlight glares if I turn around ... Any tips on where they should be? I do recall seeing them too but don't remember where :P

**Comment by combo-pizza** [comment]
Play midtown expressway then let the camera pan around your car at the beginning of the race. It should show the headlight lense flairs

**Comment by combo-pizza** [comment]
Just got the lastest git version and it had some kind of failure at the beginning of sunset drive. Gotta check again

**Comment by hrydgard** [comment]
Ah I get it, it's on night courses they show up. And indeed, on the real hardware, pretty big highlights. Hope they're not based on depth buffer readbacks, let's see.

**Comment by hrydgard** [comment]
Since the digital speedometers are OK now, I'm gonna close this, and open a new issue for the headlight glares. Something very strange seems to be going on with them.

**Comment by hrydgard** [comment]
Moved lens flare investigation to #16083 .

**Comment by combo-pizza** [comment]
Not sure where to put this but I keep getting this now. I just ended up switching to direct3d 11 because of the weird issues with vulkan in windows 11. Sun shafts still produce that odd corruption too.
![image](https://user-images.githubusercontent.com/95991264/191885679-9004a260-1413-43ad-a06a-86088bdc9294.png)

**Comment by hrydgard** [comment]
@combo-pizza did you somehow post and delete something here? I got an email notification where you report an error that I'm very surprised you got, and would like to know how to reproduce (the "readyForFence" assertion).

**Comment by stuken** [comment]
Unforunatly there is still the issue with the outline on some speedos while upscaling.
First is x4, second is x1. The same issue occurs on all rendering backends.
![UCES00422_00002](https://user-images.githubusercontent.com/8155719/191935381-e3e745e0-aa5a-42cc-a90a-9551697fe36f.jpg)
![UCES00422_00003](https://user-images.githubusercontent.com/8155719/191935385-80ccca41-de02-4b75-8409-fa052e033d90.jpg)

**Comment by hrydgard** [comment]
OK, I'll look into that. What car is that?

**Comment by Back2Life888** [comment]
It's the Gnade Esperanza

**Comment by Back2Life888** [comment]
Hmm weird. It doesn't occur on me. I have Hybrid and Bicubic 5x upscaling, Auto 1:1, Auto Max Quality, 16x Filtering and the Esperanza's speedometer works fine, hell it's even upscaled properly with no bugs and it looks crisper and sharper.

**Comment by Back2Life888** [comment]
Hey stuken, do you happen to have 4x upscaling on whilst the rest of the other settings on low?

**Comment by stuken** [comment]
All texture scaling modes dont make a difference. It disappears at auto 1:1 res scaling but appears at any of the x2, x3 and so on modes.

**Comment by Back2Life888** [comment]
Well no wonder it doesn't occur on me. Setting it on 2x, it occurs for me now but the gauge of the RR2 Crinale isn't that bad. It just has a box shadow on it like the Assoluto Bisonte (The Bisonte isn't CLUT based but it has a weird box shadow on it) and it doesn't look that major like in the Esperanza.

**Comment by hrydgard** [comment]
Alright, not so serious but of course I'll fix it anyway, looking at something else right now but will get to it relatively soon.

**Comment by combo-pizza** [comment]
@hrydgard  Sorry about that, I got frustrated because I couldn't reproduce it and just gave up. But here's the image I took from the post before
![191885679-9004a260-1413-43ad-a06a-86088bdc9294](https://user-images.githubusercontent.com/95991264/191999432-3e50ca35-02f8-49a5-8c60-2df33404592d.png)

**Comment by hrydgard** [comment]
Thanks, I'll get that fixed soon anyway, gonna rework a bit of how the Vulkan backend synchronizes.

**Comment by combo-pizza** [comment]
Thank you! I don't know if it's fixed but I at least tried opening it over fifty times with your new build and didn't get any messages like that.

**Comment by Back2Life888** [comment]
How about opening a separate issue for the garbagy sun flares?

**Comment by hrydgard** [comment]
Oh, they still happen? I really need a sure-fire way to reproduce them.

**Comment by hrydgard** [comment]
Regarding the speed it's the Galaga Esperanza, not Gnade, took me a while to find it :) 

Anyway, hm, I'm getting a much fainter effect but I can reproduce the problem. Will check it out.
